patina_dxe_core/src/locks/interrupt_mutex.rs

```diff
@@ -0,0 +1,228 @@
+//! `InterruptGuard` provides access to the protected data via dereferencing.
+//! When the `InterruptGuard` goes out of scope, the lock is released.
+//!
+//! Interrupts are disabled when the lock is held, and restored to their
+//! previous state when the lock is released.
+//!
+//! ## License
+//!
+//! Copyright (c) Microsoft Corporation.
+//!
+//! SPDX-License-Identifier: Apache-2.0
+//!
+use core::{
+    cell::UnsafeCell,
+    fmt,
+    ops::{Deref, DerefMut},
+};
+
+pub struct InterruptMutex<T: ?Sized> {
+    locked: UnsafeCell<bool>,
+    name: &'static str,
+    data: UnsafeCell<T>,
+}
+
+pub struct InterruptGuard<'a, T: ?Sized + 'a> {
+    lock: &'a UnsafeCell<bool>,
+    interrupts_enabled: bool,
+    #[allow(unused)]
+    name: &'static str,
+    data: *mut T,
+}
+
+// Safety: interrupt control is used to protect against concurrent access.
+unsafe impl<T: ?Sized + Send> Sync for InterruptMutex<T> {}
+// Safety: interrupt control is used to protect against concurrent access.
+unsafe impl<T: ?Sized + Send> Send for InterruptMutex<T> {}
+
+// Safety: interrupt control is used to protect against concurrent access.
+unsafe impl<T: ?Sized + Sync> Sync for InterruptGuard<'_, T> {}
+// Safety: interrupt control is used to protect against concurrent access.
+unsafe impl<T: ?Sized + Send> Send for InterruptGuard<'_, T> {}
+
+impl<T> InterruptMutex<T> {
+    pub const fn new(data: T, name: &'static str) -> Self {
+        Self { locked: UnsafeCell::new(false), name, data: UnsafeCell::new(data) }
+    }
+}
+
+// Interrupt control functions. Broken out to allow for mocking in tests.
+#[cfg(not(test))]
+mod interrupt_control {
+    pub fn get_interrupt_state() -> bool {
+        patina_internal_cpu::interrupts::get_interrupt_state().expect("Interrupts non-functional.")
+    }
+    pub fn disable_interrupts() {
+        patina_internal_cpu::interrupts::disable_interrupts()
+    }
+
+    pub fn restore_interrupts(state: bool) {
+        if state {
+            patina_internal_cpu::interrupts::enable_interrupts()
+        }
+    }
+}
+#[cfg(test)]
+mod interrupt_control {
+    pub fn get_interrupt_state() -> bool {
+        false
+    }
+    pub fn disable_interrupts() {}
+    pub fn restore_interrupts(_state: bool) {}
+}
+
+impl<T: ?Sized> InterruptMutex<T> {
+    /// Lock the InterruptMutex and return a InterruptGuard object used to access the data.
+    ///
+    /// Safety: Lock reentrance is not supported; attempt to re-lock something already locked will panic.
+    pub fn lock(&self) -> InterruptGuard<'_, T> {
+        self.try_lock().unwrap_or_else(|| panic!("Re-entrant locks for {:?} not permitted.", self.name))
+    }
+
+    /// Attempts to lock the InterruptMutex, and if successful, returns a guard object that can be used to access the data.
+    pub fn try_lock(&self) -> Option<InterruptGuard<'_, T>> {
+        let interrupts_enabled = interrupt_control::get_interrupt_state();
+        interrupt_control::disable_interrupts();
```

Comment:
I'm a little paranoid about the compiler doing odd stuff. We have no guarantee that that it wont reorder the interrupt and volatile write do we? Perhaps compiler barriers can help?